### DISCUSSION TITLE
Bug in scanner

### DISCUSSION THREAD

**Discussion by FateDecider** [qcha41/usit#28] [General]
Let me explain the context. Suppose I have the plot on the right of Scanner to be 'transformed' (say FFT), and I acquire data at the same time. This is fine and the plot will update the new curve. However, during this same scan, when I switch back to 'untransformed' and continue to acquire data, the 'transform' feature will glitch out and I cannot press 'transform' unless I restart the entire GUI.

**Comment by Python-simulation** [comment]
I can't reproduce the error. Can you explain in a little more detail how to reproduce it? But, from what you've said, it's most likely a pyqtgraph problem. I haven't touched fft or any of pyqtgraph's default functions. If you have to restart the GUI (I think you mean just the scanner), that means you should have errors displayed in the console. Can you also provide these?

**Comment by Python-simulation** [comment]
Hi @FateDecider, is the bug still occurring for you? Could you provide detail on the bug so I can try to reproduce it?

**Reply by Python-simulation** [reply]
I reproduced the error by using FFT before starting a scan. It raise an error in pyqtgraph when trying to do FFT on data with one point. I can fixe this by modifying pyqtgraph but I still need to see how to fixe this in autolab (by sub-classing PlotDataItem).
In the mean time you can fixe it by adding `if len(x) == 1: return np.array([0]), abs(y)` at the beginning of _fourierTransform in pyqtgraph\graphicsItems\PlotDataItem.py

**Reply by FateDecider** [reply]
Hi, thanks for checking in. We have changed our measurement instrumentation/workflow such that we do not need to scan along the time axis, so it is a bit inconvenient to try to reproduce this; but I believe the problem is exactly what you have reproduced.

**Reply by FateDecider** [reply]
Btw I have got this new error that is most likely related to starting a new scan when FFT mode is switched on. Here, the scan variable is not the same as the FFT variable. It is definitely a plotting bug:
Traceback with variables (most recent call last):
  File "C:\Users\THz-m\AppData\Local\Programs\Python\Python310\lib\site-packages\pyqtgraph\graphicsItems\PlotItem\PlotItem.py", line 943, in updateSpectrumMode
    c.setFftMode(b)
      self = <pyqtgraph.graphicsItems.PlotItem.PlotItem.PlotItem object at 0x00000275EAD5BB50>
      b = True
      c = <pyqtgraph.graphicsItems.PlotDataItem.PlotDataItem object at 0x00000275B090E680>
  File "C:\Users\THz-m\AppData\Local\Programs\Python\Python310\lib\site-packages\pyqtgraph\graphicsItems\PlotDataItem.py", line 426, in setFftMode
    self.updateItems(styleUpdate=False)
      self = <pyqtgraph.graphicsItems.PlotDataItem.PlotDataItem object at 0x00000275B090E680>
      state = True
  File "C:\Users\THz-m\AppData\Local\Programs\Python\Python310\lib\site-packages\pyqtgraph\graphicsItems\PlotDataItem.py", line 865, in updateItems
    dataset = self._getDisplayDataset()
      self = <pyqtgraph.graphicsItems.PlotDataItem.PlotDataItem object at 0x00000275B090E680>
      styleUpdate = True
      curveArgs = {'pen': 'k', 'shadowPen': None, 'fillLevel': None, 'fillOutline': False, 'brush': None, 'antialias': False, 'connect': 'auto', 'stepMode': None, 'skipFiniteCheck': False}
      scatterArgs = {'pen': 'k', 'brush': 'k', 'symbol': 'x', 'size': 10, 'data': None, 'pxMode': True, 'antialias': False, 'useCache': True}
      k = 'useCache'
      v = 'useCache'
  File "C:\Users\THz-m\AppData\Local\Programs\Python\Python310\lib\site-packages\pyqtgraph\graphicsItems\PlotDataItem.py", line 940, in _getDisplayDataset
    x,y = self._fourierTransform(x, y)
      self = <pyqtgraph.graphicsItems.PlotDataItem.PlotDataItem object at 0x00000275B090E680>
      x = array([78.07188])
      y = array([78.07188])
  File "C:\Users\THz-m\AppData\Local\Programs\Python\Python310\lib\site-packages\pyqtgraph\graphicsItems\PlotDataItem.py", line 1188, in _fourierTransform
    uniform = not np.any(np.abs(dx-dx[0]) > (abs(dx[0]) / 1000.))
      self = <pyqtgraph.graphicsItems.PlotDataItem.PlotDataItem object at 0x00000275B090E680>
      x = array([78.07188])
      y = array([78.07188])
      dx = array([], dtype=float64)
builtins.IndexError: index 0 is out of bounds for axis 0 with size 0

**Reply by Python-simulation** [reply]
This is indeed the error I spotted and discussed [here](https://github.com/pyqtgraph/pyqtgraph/issues/3018#issuecomment-2090780184).

**Comment by Python-simulation** [comment]
This bug has been fixed in [pyqtgraph](https://github.com/pyqtgraph/pyqtgraph/issues/3018#issuecomment-2090780184) and for previous pyqtgraph version, I added a fix in 01e4e67
